README.md
# polkadot

This is a package for the Blockchain Package Manager by [Blockdaemon](https://blockdaemon.com/). Deploy, maintain, and upgrade blockchain nodes on your own infrastructure.

Further reading:

* End-user documentation: https://cli.bpm.docs.blockdaemon.com/
* Developer documentation: https://sdk.bpm.docs.blockdaemon.com/

# Contributing

Pleaes use [conventional commits](https://www.conventionalcommits.org) for you commit messages. This will help us in the future to auto-generate changelogs.

New features should be developed in a branch and merged after a code review.

# Building from source

## Requirements

Make sure you have the following tools:

- [Go](https://golang.org/) is the main pogramming language. It needs to be installed
- [goreleaser](https://goreleaser.com/) is used to build binary packages. It needs to be installed
- [golangci-lint](https://github.com/golangci/golangci-lint) is used to do static code checks. It needs to be installed
- [GPG](https://gnupg.org/) is used to sign build artifacts. It needs to be installed

## Building during development

For quick development builds for your own system run:

    make build

To install the development package run:

    bpm packages install --from-file ./bin/bpm-polkadot-development

## Building a test release:

To build a test release for multiple platforms without publishing it run:

    make test-release

## Releasing a new version

You need the GPG key to sign the releases imported into your GPG keyring.

    make VERSION=<version> GITLAB_TOKEN=<gitlab-token> release

`<version>` needs to be a valid [semantic version](https://semver.org/). Do **not prefix with `v`**, the script does that automatically.

`<gitlab-token>` needs to be a gitlab token with api scope. You can create a token here: https://gitlab.com/profile/personal_access_tokens.

The artifacts will be published in Gitlab. Contact support@blockdaemon.com to add your package to the Blockchain Package Registry.

CHANGELOG.md
# 1.3.0


### Bug Fixes

* validate parameters ([0cb6d1a](https://gitlab.com/Blockdaemon/bpm-polkadot/commit/0cb6d1a327499da78fb20a05b767f406c29146ae))


### Features

* write blockchain data to disk and support for `--data-dir` ([48584ae](https://gitlab.com/Blockdaemon/bpm-polkadot/commit/48584ae776af092c9cdd73bcc9d5f4f7f955e386))






cmd/tests.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"time"
)

const (
	baseUrl       = "http://127.0.0.1:9933"
	sysChain      = "Alexander"
	sysName       = "parity-polkadot"
	tokenDecimals = 15
	tokenSymbol   = "DOT"
	sysVersion    = "0.4.4"
)

type testRunner struct {
	failed    int
	succeeded int
}

func (t *testRunner) test(testFunc func() error) {
	if err := testFunc(); err != nil {
		t.failed++
	} else {
		t.succeeded++
	}
}

func runAllTests() error {
	runner := testRunner{}

	runner.test(chainGetBlock)
	runner.test(chainGetBlockHash)
	runner.test(chainGetFinalizedHead)
	runner.test(chainGetHeader)
	runner.test(systemChain)
	runner.test(systemHealth)
	runner.test(systemName)
	runner.test(systemNetworkState)
	runner.test(systemPeers)
	runner.test(systemProperties)
	runner.test(systemVersion)

	if runner.failed > 0 {
		return fmt.Errorf("%d out of %d tests failed", runner.failed, runner.failed+runner.succeeded)
	}

	return nil
}

func chainGetBlock() error {
	// hash parameter can be set
	// only checking for no error 200 right now
	method, statusCode, expectedID, data := polkadotPost("chain_getBlock", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if statusCode == 200 {
		logMessage := fmt.Sprintf("PASSED: %s %d", method, statusCode)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func chainGetBlockHash() error {
	// blockNumber parameter can be set
	// only checking for no error 200 right now
	method, statusCode, expectedID, data := polkadotPost("chain_getBlockHash", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if statusCode == 200 {
		logMessage := fmt.Sprintf("PASSED: %s %d", method, statusCode)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func chainGetFinalizedHead() error {
	// only checking for no error 200 right now
	method, statusCode, expectedID, data := polkadotPost("chain_getFinalizedHead", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if statusCode == 200 {
		logMessage := fmt.Sprintf("PASSED: %s %d", method, statusCode)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func chainGetHeader() error {
	// hash parameter can be set
	// only checking for no error 200 right now
	method, statusCode, expectedID, data := polkadotPost("chain_getHeader", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if statusCode == 200 {
		logMessage := fmt.Sprintf("PASSED: %s %d", method, statusCode)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemChain() error {
	name := sysChain
	method, statusCode, expectedID, data := polkadotPost("system_chain", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if result == name {
		logMessage := fmt.Sprintf("PASSED: %s %s", method, name)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemHealth() error {
	method, statusCode, expectedID, data := polkadotPost("system_health", "")

	errorValue, errorKey := data["error"]
	resultMap, _ := data["result"].(map[string]interface{})
	resultIsSyncing, _ := resultMap["isSyncing"].(bool)
	resultPeers, _ := resultMap["peers"].(float64)
	resultShouldHavePeers, _ := resultMap["shouldHavePeers"].(bool)
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultIsSyncing && resultPeers > 0 && resultShouldHavePeers {
		//might need to change this if we ever expect not to have peers
		logMessage := fmt.Sprintf("PASSED: %s Syncing: %v Peers: %d Should Have Peers: %v", method, resultIsSyncing, int(resultPeers), resultShouldHavePeers)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, resultMap)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemName() error {
	name := sysName
	method, statusCode, expectedID, data := polkadotPost("system_name", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if result == name {
		logMessage := fmt.Sprintf("PASSED: %s %s", method, result)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemNetworkState() error {
	method, statusCode, expectedID, data := polkadotPost("system_networkState", "")

	errorValue, errorKey := data["error"]
	resultMap := data["result"].(map[string]interface{})
	resultAverageDownloadPerSec, _ := resultMap["averageDownloadPerSec"].(float64)
	resultAverageUploadPerSec, _ := resultMap["averageUploadPerSec"].(float64)
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultAverageDownloadPerSec > 0 && resultAverageUploadPerSec > 0 {
		logMessage := fmt.Sprintf("PASSED: %s average DL/UL %d/%d", method, int(resultAverageDownloadPerSec), int(resultAverageUploadPerSec))
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, resultMap)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemPeers() error {
	method, statusCode, expectedID, data := polkadotPost("system_peers", "")

	errorValue, errorKey := data["error"]
	result := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if len(result.([]interface{})) > 0 {
		logMessage := fmt.Sprintf("PASSED: %s %d peer(s)", method, len(result.([]interface{})))
		//logMessage := fmt.Sprintf("PASSED: %s returns peers %v", method, result)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure %v", method, result)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemProperties() error {
	method, statusCode, expectedID, data := polkadotPost("system_properties", "")

	errorValue, errorKey := data["error"]
	resultMap := data["result"].(map[string]interface{})
	resultTokenDecimals := resultMap["tokenDecimals"]
	resultTokenSymbol := resultMap["tokenSymbol"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultTokenDecimals != float64(tokenDecimals) {
		logMessage := fmt.Sprintf("FAILED: %s result decimals: %v but expected: %d", method, resultTokenDecimals, tokenDecimals)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultTokenSymbol != tokenSymbol {
		logMessage := fmt.Sprintf("FAILED: %s result symbol: %v but expected: %s", method, resultTokenSymbol, tokenSymbol)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultTokenDecimals == float64(tokenDecimals) && resultTokenSymbol == tokenSymbol {
		logMessage := fmt.Sprintf("PASSED: %s token decimals: %v token symbol: %v", method, resultTokenDecimals, resultTokenSymbol)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure", method)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func systemVersion() error {
	expectedVersion := sysVersion
	method, statusCode, expectedID, data := polkadotPost("system_version", "")

	errorValue, errorKey := data["error"]
	resultVersion := data["result"]
	messageIDValue := data["id"]

	if statusCode != 200 {
		logMessage := fmt.Sprintf("FAILED: %s status code: %d", method, statusCode)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if errorKey {
		logMessage := fmt.Sprintf("FAILED: %s error in body: %v", method, errorValue)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if messageIDValue != float64(expectedID) {
		logMessage := fmt.Sprintf("FAILED: %s message id: %f but expected %d", method, messageIDValue, expectedID)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultVersion != expectedVersion {
		logMessage := fmt.Sprintf("FAILED: %s version: %s but expected %v", method, expectedVersion, resultVersion)
		log.Print(logMessage)
		return errors.New(logMessage)
	} else if resultVersion == expectedVersion {
		logMessage := fmt.Sprintf("PASSED: %s %v", method, resultVersion)
		log.Print(logMessage)
		return nil
	} else {
		logMessage := fmt.Sprintf("FAILED: %s unknown failure", method)
		log.Print(logMessage)
		return errors.New(logMessage)
	}
}

func polkadotPost(method string, params string) (string, int, int, map[string]interface{}) {
	rand.Seed(time.Now().UnixNano())
	var messageID = rand.Int()
	requestBody, err := json.Marshal(map[string]interface{}{
		"method":  method,
		"id":      messageID,
		"jsonrpc": "2.0",
	})
	if err != nil {
		log.Fatalln(err)
	}

	resp, err := http.Post(baseUrl+params, "application/json", bytes.NewBuffer(requestBody))
	if err != nil {
		log.Fatalln(err)
	}

	defer resp.Body.Close()

	var data map[string]interface{}
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		log.Fatalln(err)
	}

	return method, resp.StatusCode, messageID, data
}


cmd/start.go
package main

import (
	"github.com/Blockdaemon/bpm-sdk/pkg/docker"
	"github.com/Blockdaemon/bpm-sdk/pkg/node"

	"context"
	"path"
	"time"

	homedir "github.com/mitchellh/go-homedir"
)

func start(currentNode node.Node) error {
	client, err := docker.NewBasicManager(currentNode.DockerPrefix(), currentNode.ConfigsDirectory())
	if err != nil {
		return err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	// First, create the docker network if it doesn't exist yet
	err = client.NetworkExists(ctx, currentNode.DockerNetworkName())
	if err != nil {
		return err
	}

	// Configure the containers
	polkadotContainer := docker.Container{
		Name:      polkadotContainerName,
		Image:     polkadotContainerImage,
		CmdFile:   polkadotCmdFile,
		NetworkID: currentNode.DockerNetworkName(),
		Mounts: []docker.Mount{
			{
				Type: "volume",
				From: polkadotDataVolumeName,
				To:   "/data",
			},
		},
		Ports: []docker.Port{
			{
				HostIP:        "0.0.0.0",
				HostPort:      "30333",
				ContainerPort: "30333",
				Protocol:      "tcp",
			},
			{
				HostIP:        "127.0.0.1",
				HostPort:      "9933",
				ContainerPort: "9933",
				Protocol:      "tcp",
			},
		},
	}

	// TODO: This is just temporarily until we have a proper authentication system we stick the certs in ~/.blockdaemon
	currentNode.Collection.Key, err = homedir.Expand(currentNode.Collection.Key)
	if err != nil {
		return err
	}
	currentNode.Collection.Cert, err = homedir.Expand(currentNode.Collection.Cert)
	if err != nil {
		return err
	}
	currentNode.Collection.CA, err = homedir.Expand(currentNode.Collection.CA)
	if err != nil {
		return err
	}
	// TODO end

	polkadotbeatContainer := docker.Container{
		Name:      polkadotbeatContainerName,
		Image:     polkadotbeatContainerImage,
		Cmd:       []string{"-e", "-strict.perms=false"},
		NetworkID: currentNode.DockerNetworkName(),
		Mounts: []docker.Mount{
			{
				Type: "bind",
				From: path.Join(currentNode.ConfigsDirectory(), polkadotbeatConfigFile),
				To:   "/usr/share/polkadotbeat/polkadotbeat.yml",
			},
			{
				Type: "bind",
				From: currentNode.Collection.CA,
				To:   "/etc/ssl/beats/ca.crt",
			},
			{
				Type: "bind",
				From: currentNode.Collection.Cert,
				To:   "/etc/ssl/beats/beat.crt",
			},
			{
				Type: "bind",
				From: currentNode.Collection.Key,
				To:   "/etc/ssl/beats/beat.key",
			},
		},
	}

	filebeatContainer := docker.Container{
		Name:      filebeatContainerName,
		Image:     filebeatContainerImage,
		Cmd:       []string{"-e", "-strict.perms=false"},
		NetworkID: currentNode.DockerNetworkName(),
		Mounts: []docker.Mount{
			{
				Type: "bind",
				From: filebeatConfigFile,
				To:   "/usr/share/filebeat/filebeat.yml",
			},
			{
				Type: "bind",
				From: "/var/lib/docker/containers",
				To:   "/var/lib/docker/containers",
			},
			{
				Type: "bind",
				From: currentNode.Collection.CA,
				To:   "/etc/ssl/beats/ca.crt",
			},
			{
				Type: "bind",
				From: currentNode.Collection.Cert,
				To:   "/etc/ssl/beats/beat.crt",
			},
			{
				Type: "bind",
				From: currentNode.Collection.Key,
				To:   "/etc/ssl/beats/beat.key",
			},
		},
		User: "root",
	}

	// Next, start the containers
	if err := client.ContainerRuns(ctx, polkadotbeatContainer); err != nil {
		return err
	}
	if err := client.ContainerRuns(ctx, filebeatContainer); err != nil {
		return err
	}

	return client.ContainerRuns(ctx, polkadotContainer)
}


cmd/status.go
package main

import (
	"github.com/Blockdaemon/bpm-sdk/pkg/docker"
	"github.com/Blockdaemon/bpm-sdk/pkg/node"

	"context"
	"time"
)

func status(currentNode node.Node) (string, error) {
	client, err := docker.NewBasicManager(currentNode.DockerPrefix(), currentNode.ConfigsDirectory())
	if err != nil {
		return "", err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	polkadotContainerRunning, err := client.IsContainerRunning(ctx, polkadotContainerName)
	if err != nil {
		return "", err
	}
	polkadotbeatContainerRunning, err := client.IsContainerRunning(ctx, polkadotbeatContainerName)
	if err != nil {
		return "", err
	}
	filebeatCotainerRunning, err := client.IsContainerRunning(ctx, filebeatContainerName)
	if err != nil {
		return "", err
	}

	if polkadotContainerRunning && polkadotbeatContainerRunning && filebeatCotainerRunning {
		return "running", nil
	} else if polkadotContainerRunning || polkadotbeatContainerRunning || filebeatCotainerRunning {
		return "incomplete", nil
	}

	return "stopped", nil
}


cmd/configs.go
package main

import (
	"github.com/Blockdaemon/bpm-sdk/pkg/node"
	"github.com/Blockdaemon/bpm-sdk/pkg/template"
)

const (
	polkadotbeatConfigTpl = `polkadotbeat:
    period: 30s
    polkadot_host: "{{ .ContainerName "polkadot" }}"
    polkadot_port: "9933"
fields:
    info:
        launch_type: bpm
        node_xid: {{ .ID }}
        project: development
        protocol_type: POLKADOT
        network_type: public
        user_id: TODO
        environment: {{ .Environment }}
fields_under_root: true
output:
    logstash:
        hosts:
        - "{{ .Collection.Host }}"
        ssl:
            certificate: /etc/ssl/beats/beat.crt
            certificate_authorities:
            - /etc/ssl/beats/ca.crt
            key: /etc/ssl/beats/beat.key
`

	filebeatConfigTpl = `filebeat.inputs:
- type: docker
  containers.ids:
  - '*'

filebeat.config:
  modules:
    path: ${path.config}/modules.d/*.yml
    reload.enabled: false

fields:
    info:
        launch_type: bpm
        node_xid: {{ .ID }}
        project: development
        protocol_type: POLKADOT
        network_type: public
        user_id: TODO
        environment: {{ .Environment }}
fields_under_root: true
output:
    logstash:
        hosts:
        - "{{ .Collection.Host }}"
        ssl:
            certificate: /etc/ssl/beats/beat.crt
            certificate_authorities:
            - /etc/ssl/beats/ca.crt
            key: /etc/ssl/beats/beat.key
`
)

func createConfigs(currentNode node.Node) error {
	if err := template.ConfigFileRendered(filebeatConfigFile, filebeatConfigTpl, currentNode); err != nil {
		return err
	}

	return template.ConfigFileRendered(polkadotbeatConfigFile, polkadotbeatConfigTpl, currentNode)
}


cmd/parameter_validator.go
package main

import (
	"fmt"

	"go.blockdaemon.com/bpm/sdk/pkg/node"
	"go.blockdaemon.com/bpm/sdk/pkg/plugin"
)

// polkadotParameterValidator validates parameters
type polkadotParameterValidator struct {
	plugin.SimpleParameterValidator
}

// ValidateParameters uses SimpleParameterValidator but also check is the network is correct
func (t polkadotParameterValidator) ValidateParameters(currentNode node.Node) error {
	// Call SimpleParameterValidator to check if all parameters actually have values
	if err := t.SimpleParameterValidator.ValidateParameters(currentNode); err != nil {
		return err
	}

	subtype := currentNode.StrParameters["subtype"]
	if subtype != "watcher" && subtype != "validator" {
		return fmt.Errorf("unknown subtype: %q", subtype)
	}

	return nil
}

// NewpolkadotParameterValidator creates a new instance of polkadotParameterValidator
func NewpolkadotParameterValidator(pluginParameters []plugin.Parameter) polkadotParameterValidator {
	return polkadotParameterValidator{
		SimpleParameterValidator: plugin.NewSimpleParameterValidator(pluginParameters),
	}
}


cmd/main.go
package main

import (
	"go.blockdaemon.com/bpm/sdk/pkg/docker"
	"go.blockdaemon.com/bpm/sdk/pkg/node"
	"go.blockdaemon.com/bpm/sdk/pkg/plugin"
)

var version string

const (
	polkadotContainerImage = "docker.io/chevdor/polkadot:0.4.4"
	polkadotContainerName  = "polkadot"
	polkadotCmdFile        = "polkadot.dockercmd"

	collectorContainerName = "collector"
	collectorImage         = "docker.io/blockdaemon/polkadot-collector:0.5.0"
	collectorEnvFile       = "configs/collector.env"
)

type PolkadotTester struct{}

func (d PolkadotTester) Test(currentNode node.Node) (bool, error) {
	if err := runAllTests(); err != nil {
		return false, err
	}
	return true, nil
}

func main() {
	templates := map[string]string{
		polkadotCmdFile:  polkadotCmdTpl,
		collectorEnvFile: collectorEnvTpl,
	}

	parameters := []plugin.Parameter{
		{
			Name:        "subtype",
			Type:        plugin.ParameterTypeString,
			Description: "The type of node. Must be either `watcher` or `validator`",
			Mandatory:   false,
			Default:     "watcher",
		},
		{
			Name:        "validator-key",
			Type:        plugin.ParameterTypeString,
			Description: "The key used for a validator (required if subtype = validator)",
			Mandatory:   false,
		},
	}

	containers := []docker.Container{
		{
			Name:    polkadotContainerName,
			Image:   polkadotContainerImage,
			CmdFile: polkadotCmdFile,
			Mounts: []docker.Mount{
				{
					Type: "bind",
					From: "{{ index .Node.StrParameters \"data-dir\" }}",
					To:   "/data",
				},
			},
			Ports: []docker.Port{
				{
					HostIP:        "0.0.0.0",
					HostPort:      "30333",
					ContainerPort: "30333",
					Protocol:      "tcp",
				},
				{
					HostIP:        "127.0.0.1",
					HostPort:      "9933",
					ContainerPort: "9933",
					Protocol:      "tcp",
				},
			},
			CollectLogs: true,
		},
		{
			Name:        collectorContainerName,
			Image:       collectorImage,
			EnvFilename: collectorEnvFile,
			Mounts: []docker.Mount{
				{
					Type: "bind",
					From: "logs",
					To:   "/data/nodestate",
				},
			},
			CollectLogs: true,
		},
	}

	description := "A polkadot package"

	polkadotPlugin := plugin.NewDockerPlugin("polkadot", version, description, parameters, templates, containers)
	polkadotPlugin.ParameterValidator = NewpolkadotParameterValidator(parameters)
	polkadotPlugin.Tester = PolkadotTester{}

	plugin.Initialize(polkadotPlugin)
}
